### PR TITLE
Replace layer dims with private named tuple

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR drops the public layer dims, which was deprecated and to be removed in `0.4.2`. It also replaces the private `Dims` object with a simple `Dims` `NamedTuple`. While this may not seen like a big switch it's actually a huge simplification in complexity and will make gradually refactoring to remove the dims information from the layer much much easier as now there is a small named tuple which can gradually become smaller, instead of a full object. It also makes refactors like #1917 as now we only have to worry about one copy of the real dims object instead of many.

It also makes it clear that what we want to remove from the layer next is really the `dims.ndisplay` concept, and that actually a version of `dims.order` concept probably deserves to be fully to be fully promoted to live on the layer to enable things like https://github.com/napari/napari/issues/1904#issuecomment-731023666. Similarly now it will be much easier to add layer specific `axis_labels`.

## Type of change

- [x] Refactor (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by sofroniewn** [comment]
@Czaki raised some concerns about the performance with `NamedTuples` in recent discussion, and we should check if there is a performance degradation here before merge

**Comment by sofroniewn** [comment]
Still have a couple things to fix on this one

**Comment by sofroniewn** [comment]
Ok, this PR is ready for review - it is marked as a breaking change as it removes the `layer.dims` property, which is set to be removed in `0.4.3` and has been deprecated since the `0.4.0` release, but otherwise it changes nothing.

It does this incrementally by moving the additional private attributes we still need to things like `layer._dims_displayed`. While this may seem still complex, it will make it much easier to continue to remove the dims information from the layer until we get to the point where it is gone. It no longer uses a `NamedTuple` as was initially done in the implementation. It was not necessary.

One note is that #1955 should be merged before this otherwise the undo/ redo on labels will be messed up. But otherwise this is good to go. Feedback from @jni and @Czaki welcome

**Comment by Czaki** [comment]
I'm a little worried. You ungroup part of variables. It will increase the complexity of this class. The layer class is quite big, so each additional variable adds a little work to maintability. What is the benefit of remove grouping?

**Comment by sofroniewn** [comment]
> I'm a little worried. You ungroup part of variables. It will increase the complexity of this class. The layer class is quite big, so each additional variable adds a little work to maintability. What is the benefit of remove grouping?

Over the next couple of releases I want to remove each of these variables until we can remove the concept of the sliced data from the layer entirely. It's going to be much easier to keep removing things now that the variables are separated and the `Dims` object itself is gone. Once all these variables are gone the layer object will be much simpler!

File: napari/layers/base/base.py [hunk 1]
```diff
@@ -486,34 +488,45 @@ def position(self, position):
...
-            (
-                "The layer.dims parameter is deprecated and will be removed in version 0.4.3."
-                " Instead you should use the viewer.dims parameter on the main viewer object."
-            ),
-            category=DeprecationWarning,
-            stacklevel=2,
-        )
-        return self._dims
+    def _dims_displayed(self):
+        """To be removed displayed dimensions."""
```
Comment (jni) [inline napari/layers/base/base.py:491]: Can you add a more elaborate note for "to be removed"? Maybe a link to discussion about this? "ultimately we aim to remove all slicing information from the layer itself, see (link)"?

**Review by jni — APPROVED** [review]
👍 just left a small note

**Comment by jni** [comment]
btw is the macOS failure related? It doesn't seem to be but I don't have a good overview of whether it's surfaced elsewhere.

File: napari/layers/base/base.py (hunk 1 above)
Comment (sofroniewn) [reply napari/layers/base/base.py:491]: added

**Comment by sofroniewn** [comment]
Comments added and updated from master. I've given this some tests and I think it's ready to go. All tests passing, will merge soon

**Pull request merged**